GetIntoTeachingApi/Jobs/LocationSyncJob.cs

```diff
@@ -99,28 +101,41 @@ private async Task SyncLocations(string csvPath)
                 batch.Add(location);
                 locationCount++;
 
-                QueueBatch(batch);
+                await ProcessBatch(batch);
             }
 
-            QueueBatch(batch, true);
+            await ProcessBatch(batch, true);
 
-            var batchJobCount = (int)Math.Ceiling((decimal)locationCount / BatchInterval);
-            _logger.LogInformation($"LocationSyncJob - Queueing {locationCount} Locations ({batchJobCount} Jobs)");
+            var batchCount = (int)Math.Ceiling((decimal)locationCount / BatchInterval);
+            _logger.LogInformation($"LocationSyncJob - Processed {locationCount} Locations ({batchCount} Batches)");
         }
 
-        private void QueueBatch(ICollection<dynamic> batch, bool force = false)
+        private async Task ProcessBatch(ICollection<Location> batch, bool force = false)
         {
             if (!force && batch.Count != BatchInterval)
             {
                 return;
-            }
-
-            // Batch is serialized to pass by value.
-            _jobClient.Enqueue<LocationBatchJob>(x => x.RunAsync(JsonConvert.SerializeObject(batch)));
+            }
+
+            var parameters = batch.SelectMany((location, index) => DbParameters(location, index));
+            var values = string.Join(", ", batch.Select((_, index) => $"(@postcode{index}, @coordinate{index}, @source{index})"));
+
+            await _dbContext.Database.ExecuteSqlRawAsync(
+                $"INSERT INTO public.\"Locations\" (\"Postcode\", \"Coordinate\", \"Source\") VALUES {values} ON CONFLICT DO NOTHING", parameters);

```

Comment:
I guess this is using raw SQL rather than something like `SqlBulkCopy` because it's PostgreSQL rather than SQL Server?

Reply:
I thought about using `COPY` as that would be lightning fast, but the issue is we can't straight ingest the CSV as we sanitize the postcode (removing white space and lower casing) and convert the lat/lng to a `Point` with the correct SRID.